qiskit/quantum_info/operators/unitary.py

```diff
@@ -0,0 +1,268 @@
+
+        Args:
+            inplace (bool): attempt inplace (relies on numpy.transpose)
+
+        Returns:
+            Unitary: transposed unitary
+        """
+        if inplace:
+            self.matrix = self.matrix.transpose()
+            return self
+        else:
+            output = copy.deepcopy(self)
+            output.matrix = self.matrix.transpose()
+            return output
+
+    @property
+    def matrix(self):
+        """
+        Get matrix representation
+
+        Returns:
+            numpy.ndarray: matrix
+
+        Raises:
+            QiskitError: if representation not defined
+        """
+        if self._matrix is not None:
+            return self._matrix
+        else:
+            raise QiskitError("matrix representation not defined")
+
+    def power(self, n, inplace=False):
+        """Return n-th matrix power.
+
+        Args:
+            n (int): integer power
+            inplace (bool): whether to do operation in place.
+
+        Returns:
+            Unitary: Unitary power
+        """
+        if inplace:
+            if n >= 0:
+                self.matrix = numpy.linalg.matrix_power(self.matrix, n)
+            else:
+                self.matrix = numpy.linalg.matrix_power(
+                    self.matrix.T.conj(), n)
+            return self
+        else:
+            dim = self.dimension
+            uni = Unitary(numpy.empty((dim, dim), dtype='complex'), validate=False)
+            if n >= 0:
+                uni.matrix = numpy.linalg.matrix_power(self.matrix, n)
+            else:
+                uni.matrix = numpy.linalg.matrix_power(
+                    self.matrix.T.conj(), n)
+            return uni
+
+    @matrix.setter
+    def matrix(self, mat):
+        """set matrix representation
+
+        Args:
+            mat (ndarray | list(list)): matrix to set
+
+        Raises:
+            QiskitError: if matrix is not unitary
+        """
+        if mat is not None:
+            mat = numpy.asarray(mat, dtype='complex')
+            if self._validate:
+                if not numpy.allclose(mat.T.conj() @ mat, numpy.identity(mat.shape[0]),
+                                      rtol=self._rtol, atol=self._atol):
+                    raise QiskitError("matrix is not unitary")
+            self._matrix = mat
+            self._n_qubits = int(numpy.log2(mat.shape[0]))
+            self._dimension = mat.shape[0]
+        else:
+            self._matrix = None  # for creating empty Unitary
+
+    def reapply(self, circ):
+        """reapply this gate to circ; TODO: remove when monkey patching removed"""
+        self._modifiers(circ._attach(self))
+
+    def _define_decompositions(self):
+        decomposition = DAGCircuit()
+        self._decompositions = [decomposition]
+
+    @property
+    def label(self):
+        """get unitary label
+
+        Returns:
+            str: object's label
+        """
+        return self._label
+
+    @label.setter
+    def label(self, name):
```

Comment:
As with `matrix` I would consider not making the `label` setter method public, and have it done by making a new object.